src/lazarus/ECS/ECSEngine.h

```
@@ -0,0 +1,203 @@
+
+        if (!includeDeleted && entity->isDeleted())
+            continue;
+
+        if (entity->has<Types...>())
+            func(entity, entity->get<Types>()...);
+    }
+}
+
+template <typename EventType>
+void ECSEngine::subscribe(EventListener<EventType>* eventListener)
+{
+    std::type_index typeId = __lz::getTypeIndex<EventType>();
+    auto found = subscribers.find(typeId);
+    if (found == subscribers.end())
+    {
+        // No subscribers to this type of event yet, create vector
+        std::vector<__lz::BaseEventListener*> vec;
+        vec.push_back(eventListener);
+        subscribers[typeId] = vec;
+    }
+    else
+    {
+        // There already exists a list of subscribers to this event type
+        found->second.push_back(eventListener);
+    }
+}
+
+template <typename EventType>
+void ECSEngine::unsubscribe(EventListener<EventType>* eventListener)
+{
+    auto found = subscribers.find(__lz::getTypeIndex<EventType>());
+    if (found != subscribers.end())
+    {
+        auto eventSubscribers = found->second;
+        for (auto it = eventSubscribers.begin(); it != eventSubscribers.end(); ++it)
+        {
+            if (*it == eventListener)
+            {
+                // System found, remove it from the subscriber list
+                eventSubscribers.erase(it);
+                return;
+            }
+        }
+    }
+    // System was not found
+    throw __lz::LazarusException("ECS engine was not subscribed to the given event");
+}
+
+template <typename EventType>
+void ECSEngine::emit(const EventType& event)
+{
+    auto found = subscribers.find(__lz::getTypeIndex<EventType>());
+    if (found != subscribers.end())
+    {
+        auto eventSubscribers = found->second;
+        for (auto it = eventSubscribers.begin(); it != eventSubscribers.end(); ++it)
+        {
+            auto* listener = dynamic_cast<EventListener<EventType>*>(*it);
+            listener->receive(*this, event);
+        }
+    }
```

Should we silently ignore EventTypes without subscribed EventListeners? Maybe we could log as debug something like "Event of type %(event)s was emitted, but no EventListener was subscribed to it!" as debug information, the rationale being that someone forgot to code the subscription of the EventListener(s)